content/en/blog/_posts/2023-04-13-hpa-container-resource-metric.md

```diff
@@ -0,0 +1,98 @@
+---
+layout: blog
+title: "HPA ContainerResource type metric moves to beta"
+date: 2023-04-11
+slug: hpa-container-resource-metric
+---
+
+**Author: [Kensei Nakada](https://github.com/sanposhiho) (Mercari)**
+
+Kubernetes 1.20 introduced [the ContainerResource type metric](/docs/tasks/run-application/horizontal-pod-autoscale/#container-resource-metrics) in HPA.
+
+In Kubernetes 1.27, this feature moves to beta and the corresponding feature gate (`HPAContainerMetrics`) gets enabled by default. 
+
+## What is the ContainerResource type metric
+
+The ContainerResource type metric allows us to configure the autoscaling based on resource usage of individual containers.
+
+In the following example, the HPA controller scales the target 
+so that the average utilization of the cpu in the application container of all the pods is around 60%.
+(See [the algorithm details](/docs/tasks/run-application/horizontal-pod-autoscale/#algorithm-details)
+to know how the desired replica number is calculated exactly)
+
+```yaml
+type: ContainerResource
+containerResource:
+  name: cpu
+  container: application
+  target:
+    type: Utilization
+    averageUtilization: 60
+```
+
+## The difference from the Resource type metric
+
+HPA already has [the Resource type metric](/docs/tasks/run-application/horizontal-pod-autoscale/#support-for-resource-metrics). 
+
+We can define the target resource utilization like the following,
+and then HPA will scale up/down the replicas based on the current utilization.
+
+```yaml
+type: Resource
+resource:
+  name: cpu
+  target:
+    type: Utilization
+    averageUtilization: 60
+```
+
+But, this Resource type metric refers to the average utilization of the **Pods**.
+
+In case a Pod has multiple containers, the utilization calculation would be:
+
+```
+sum{the resource usage of each container} / sum{the resource request of each container}
+```
+
+The resource utiliaztion of each container may not have a direct correlation or may grow at different rates as the load changes. 
+
+For example:
+- A sidecar container is only providing an auxiliary service such as log shipping. If the application does not log very frequently or does not produce logs in its hotpath then the usage of the log shipper will not grow.
+- A sidecar container which provides authentication. Due to heavy caching the usage will only increase slightly when the load on the main container increases. In the current blended usage calculation approach this usually results in the the HPA not scaling up the deployment because the blended usage is still low.
+- A sidecar may be injected without resources set which prevents scaling based on utilization. In the current logic the HPA controller can only scale on absolute resource usage of the pod when the resource requests are not set.
```

Comment:
```suggestion
- A sidecar container is only providing an auxiliary service such as log shipping.
  If the application does not log very frequently or does not produce logs in its hotpath
  then the usage of the log shipper will not grow.
- A sidecar container which provides authentication. Due to heavy caching
  the usage will only increase slightly when the load on the main container increases.
  In the current blended usage calculation approach this usually results in
  the HPA not scaling up the deployment because the blended usage is still low.
- A sidecar may be injected without resources set which prevents scaling
  based on utilization. In the current logic the HPA controller can only scale
  on absolute resource usage of the pod when the resource requests are not set.
```